### DISCUSSION TITLE
Image field title/filename options

### DISCUSSION THREAD

**Discussion by EternalBlack** [engram-design/FeedMe#895] [Ideas]
### Description

Having an image (asset-)field to map feed content to while having no control over the resulting image is quite difficult. Especially if you have multiple feeds with the same image name coming from the feed.

### Steps to reproduce

1. Create a feed that maps to a section where this section has an asset-field (image).
2. Run feed.
3. Realize just because inside the feed the file-URL has 4000 characters your image-title and name now too is 4000 characters.

Other scenario
1. Create a feed that maps to a section where this section has an asset-field (image).
2. Create another feed that maps to a section where this section has an asset-field (image).
3. Run both feeds.
4. First feed has `www.mydomain.com/image.jpg` as image, second feed has `www.yourdomain.com/image.jpg`. This would result in the later overwriting the first image, just because it has the same name.

### Additional info

Would be nice to have the option to alter the image name (or on a per property basis for this available fields of the asset) via a text-field that accepts Twig. So something like this would be possible:

| Field | Command |
| --- | --- |
| Title | {{ feedID }}{{ imageTitle\|hash\|raw }} |
| Filename | {{ feedID }}{{ imageTitle\|hash\|raw }}{{ fileExtension }} |

- Plugin version: 4.0.0 (former `pro` user...)
- Craft version: 3.1.22 Pro

**Comment by EternalBlack** [comment]
First of all: I'm glad that this is considered an enhancement!
Second... This is also kind of related to #355 as there we got the problem with duplicate images due to `Feed Me` thinking that the image is no there yet (as craft reports back that it isn't but once saved it's like "Oh, we already got one here, lets just save this one too with the current timestamp appended"...).

As this is kind of urgent to us (we got ~600Feeds where 300 of them import images and we need to re-run those feeds multiple times daily... resulting in having many unused duplicates that need to be cleaned up every time). Is there any option to interfere with the image saving process of Feed-Me? I.e. an [Event](https://docs.craftcms.com/feed-me/v4/developers/events.html#feed-processing-related-events) that I could potentially hook into to alter the image name based on the `feed-id`? As far as I can tell there is none yet, but maybe I'm overlooking something... 🤔

**Comment by brandonkelly** [comment]
As of the next release, plugins/module will be able to modify the parsed values of each field, via the `onAfterParseField` event on `craft\feedme\services\Fields`.

```php
use yii\base\Event;
use craft\feedme\services\Fields;
use craft\feedme\events\FieldEvent;

Event::on(Fields::class, Fields::EVENT_AFTER_PARSE_FIELD, function(FieldEvent $event) {
    if ($event->fieldHandle === 'myAssetsField' && $event->fieldValue !== null) {
        // $event->fieldValue will be an array of assets (already saved).
        // Modify the assets or the field value as needed...
    }
});
```

To get the fix early. change your `craftcms/feed-me` requirement in composer.json to:

```json
"require": {
  "craftcms/feed-me": "dev-develop#272c933b11a3bad21a77fe611c79b47312d5fe91 as 4.1.0",
  "...": "..."
}
```

Then run `composer update`.

**Comment by kristiansp** [comment]
@brandonkelly Thanks for this. Could it be an idea to also let other events (notably the `onBeforeParseField`) also modify the data passed to the event?

I have a feed where I _know_ that I need to do some modification on the values imported to a multiselect field (for instance, there are some different values basically meaning the same, that I would need to set to the value I use in the multiselect). I can use the above addition, but at least for values that are not matching any options in the multiselect, I have an empty `$event->parsedValue`, and thus need to go back to the `$event->feedData` and fetch it by the name it has in the feed.

Since I know I'm going to have to do this anyways, it would be much cleaner to do it _before_ FeedMe tries to set it to the model, instead of after it predictably fails to do so, and then I could just check for the feed handle and do my thing to the values.

**Comment by jamesmacwhite** [comment]
Please someone let me know if I'm being dumb, but $event->fieldValue, is throwing an unknown property exception, it doesn't appear to exist:

![image](https://user-images.githubusercontent.com/8067792/74947184-6894e180-53f2-11ea-88c5-09350689e916.png)

Feed Me version: 4.2.0.1

**Comment by thijskaspers** [comment]
> onAfterParseField

@brandonkelly I was looking for an event like `onAfterParseField`. Just so you know, it is missing from the feedme docs, found out about it here.
